// @solana/wallet-adapter-vue/lib/cjs/useWallet.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createWalletStore = exports.initWallet = exports.provideWallet = exports.useWallet = void 0;
const wallet_adapter_base_1 = require("@solana/wallet-adapter-base");
const runtime_core_1 = require("@vue/runtime-core");
const errors_1 = require("./errors");
const useLocalStorage_1 = require("./useLocalStorage");
const walletStoreKey = Symbol();
let globalWalletStore = null;
const useWallet = () => {
    const providedWalletStore = (0, runtime_core_1.inject)(walletStoreKey, undefined);
    if (providedWalletStore)
        return providedWalletStore;
    if (globalWalletStore)
        return globalWalletStore;
    throw new Error('Wallet not initialized. Please use the `WalletProvider` component to initialize the wallet.');
};
exports.useWallet = useWallet;
const provideWallet = (walletStoreProps) => {
    (0, runtime_core_1.provide)(walletStoreKey, (0, exports.createWalletStore)(walletStoreProps));
};
exports.provideWallet = provideWallet;
const initWallet = (walletStoreProps) => {
    globalWalletStore = (0, exports.createWalletStore)(walletStoreProps);
};
exports.initWallet = initWallet;
const createWalletStore = ({ wallets, autoConnect = false, onError = (error) => console.error(error), localStorageKey = 'walletName', }) => {
    const name = (0, useLocalStorage_1.useLocalStorage)(localStorageKey);
    const wallet = (0, runtime_core_1.ref)(null);
    const adapter = (0, runtime_core_1.ref)(null);
    const publicKey = (0, runtime_core_1.ref)(null);
    const ready = (0, runtime_core_1.ref)(false);
    const connected = (0, runtime_core_1.ref)(false);
    const connecting = (0, runtime_core_1.ref)(false);
    const disconnecting = (0, runtime_core_1.ref)(false);
    // Helper methods to set and reset the main state variables.
    const setState = (state) => {
        wallet.value = state.wallet;
        adapter.value = state.adapter;
        ready.value = state.ready;
        publicKey.value = state.publicKey;
        connected.value = state.connected;
    };
    const resetState = () => {
        setState({
            wallet: null,
            adapter: null,
            ready: false,
            publicKey: null,
            connected: false,
        });
    };
    // Create a wallet dictionary keyed by their name.
    const walletsByName = (0, runtime_core_1.computed)(() => {
        return wallets.reduce((walletsByName, wallet) => {
            walletsByName[wallet.name] = wallet;
            return walletsByName;
        }, {});
    });
    // Update the wallet and adapter based on the wallet provider.
    (0, runtime_core_1.watch)(name, () => {
        var _a, _b;
        const wallet = (_b = (_a = walletsByName.value) === null || _a === void 0 ? void 0 : _a[name.value]) !== null && _b !== void 0 ? _b : null;
        const adapter = wallet && wallet.adapter;
        if (adapter) {
            setState({
                wallet,
                adapter,
                ready: false,
                publicKey: adapter.publicKey,
                connected: adapter.connected,
            });
            // Asynchronously update the ready state
            const waiting = name;
            (function () {
                return __awaiter(this, void 0, void 0, function* () {
                    const readyValue = yield adapter.ready();
                    // If the selected wallet hasn't changed while waiting, update the ready state
                    if (name === waiting) {
                        ready.value = readyValue;
                    }
                });
            })();
        }
        else {
            resetState();
        }
    }, { immediate: true });
    // Select a wallet by name.
    const select = (walletName) => __awaiter(void 0, void 0, void 0, function* () {
        if (name.value === walletName)
            return;
        if (adapter.value)
            yield adapter.value.disconnect();
        name.value = walletName;
    });
    // Handle the adapter events.
    const onDisconnect = () => (name.value = null);
    const onConnect = () => {
        if (!adapter.value)
            return;
        publicKey.value = adapter.value.publicKey;
        connected.value = adapter.value.connected;
    };
    const invalidateListeners = (0, runtime_core_1.watchEffect)((onInvalidate) => {
        const _adapter = adapter.value;
        if (!_adapter)
            return;
        _adapter.on('connect', onConnect);
        _adapter.on('disconnect', onDisconnect);
        _adapter.on('error', onError);
        onInvalidate(() => {
            _adapter.off('connect', onConnect);
            _adapter.off('disconnect', onDisconnect);
            _adapter.off('error', onError);
        });
    });
    if (typeof window !== 'undefined') {
        // Ensure the adapter listeners are invalidated before refreshing the page.
        // This is because Vue does not unmount components when the page is being refreshed.
        window.addEventListener('beforeunload', invalidateListeners);
    }
    // Helper method to return an error whilst using the onError callback.
    const newError = (error) => {
        onError(error);
        return error;
    };
    // Connect the adapter to the wallet.
    const connect = () => __awaiter(void 0, void 0, void 0, function* () {
        if (connected.value || connecting.value || disconnecting.value)
            return;
        if (!wallet.value || !adapter.value)
            throw newError(new errors_1.WalletNotSelectedError());
        if (!ready.value) {
            name.value = null;
            if (typeof window !== 'undefined') {
                window.open(wallet.value.url, '_blank');
            }
            throw newError(new wallet_adapter_base_1.WalletNotReadyError());
        }
        try {
            connecting.value = true;
            yield adapter.value.connect();
        }
        catch (error) {
            name.value = null;
            throw error;
        }
        finally {
            connecting.value = false;
        }
    });
    // Disconnect the adapter from the wallet.
    const disconnect = () => __awaiter(void 0, void 0, void 0, function* () {
        if (disconnecting.value)
            return;
        if (!adapter.value) {
            name.value = null;
            return;
        }
        try {
            disconnecting.value = true;
            yield adapter.value.disconnect();
        }
        finally {
            name.value = null;
            disconnecting.value = false;
        }
    });
    // Send a transaction using the provided connection.
    const sendTransaction = (transaction, connection, options) => __awaiter(void 0, void 0, void 0, function* () {
        if (!adapter.value)
            throw newError(new errors_1.WalletNotSelectedError());
        if (!connected.value)
            throw newError(new wallet_adapter_base_1.WalletNotConnectedError());
        return yield adapter.value.sendTransaction(transaction, connection, options);
    });
    // Sign a transaction if the wallet supports it.
    const signTransaction = (0, runtime_core_1.computed)(() => {
        const _adapter = adapter.value;
        if (!(_adapter && 'signTransaction' in _adapter))
            return;
        return (transaction) => __awaiter(void 0, void 0, void 0, function* () {
            if (!connected.value)
                throw newError(new wallet_adapter_base_1.WalletNotConnectedError());
            return yield _adapter.signTransaction(transaction);
        });
    });
    // Sign multiple transactions if the wallet supports it
    const signAllTransactions = (0, runtime_core_1.computed)(() => {
        const _adapter = adapter.value;
        if (!(_adapter && 'signAllTransactions' in _adapter))
            return;
        return (transactions) => __awaiter(void 0, void 0, void 0, function* () {
            if (!connected.value)
                throw newError(new wallet_adapter_base_1.WalletNotConnectedError());
            return yield _adapter.signAllTransactions(transactions);
        });
    });
    // Sign an arbitrary message if the wallet supports it.
    const signMessage = (0, runtime_core_1.computed)(() => {
        const _adapter = adapter.value;
        if (!(_adapter && 'signMessage' in _adapter))
            return;
        return (message) => __awaiter(void 0, void 0, void 0, function* () {
            if (!connected.value)
                throw newError(new wallet_adapter_base_1.WalletNotConnectedError());
            return yield _adapter.signMessage(message);
        });
    });
    // If autoConnect is enabled, try to connect when the adapter changes and is ready.
    (0, runtime_core_1.watchEffect)(() => __awaiter(void 0, void 0, void 0, function* () {
        if (!autoConnect || !adapter.value || !ready.value || connected.value || connecting.value)
            return;
        try {
            connecting.value = true;
            yield adapter.value.connect();
        }
        catch (error) {
            // Clear the selected wallet
            name.value = null;
            // Don't throw error, but onError will still be called
        }
        finally {
            connecting.value = false;
        }
    }));
    // Return the created store.
    return {
        // Props.
        wallets,
        autoConnect,
        // Data.
        wallet,
        adapter,
        publicKey,
        ready,
        connected,
        connecting,
        disconnecting,
        // Methods.
        select,
        connect,
        disconnect,
        sendTransaction,
        signTransaction,
        signAllTransactions,
        signMessage,
    };
};
exports.createWalletStore = createWalletStore;
