// @solana/wallet-adapter-vue/lib/cjs/WalletProvider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WalletProvider = void 0;
const runtime_core_1 = require("@vue/runtime-core");
const useWallet_1 = require("./useWallet");
exports.WalletProvider = (0, runtime_core_1.defineComponent)({
    name: 'wallet-provider',
    props: {
        wallets: {
            type: Array,
            default: () => [],
        },
        autoConnect: {
            type: Boolean,
            default: false,
        },
        onError: {
            type: Function,
            default: (error) => console.error(error),
        },
        localStorageKey: {
            type: String,
            default: 'walletName',
        },
    },
    setup(props, { slots }) {
        (0, useWallet_1.provideWallet)({
            wallets: props.wallets,
            autoConnect: props.autoConnect,
            onError: props.onError,
            localStorageKey: props.localStorageKey,
        });
        return () => { var _a; return (_a = slots.default) === null || _a === void 0 ? void 0 : _a.call(slots); };
    },
});
